### PR TITLE
toying with Docker for testing

### PR DESCRIPTION
**Pull request by JBKahn**
I was toying around with using docker to test this locally since I don't know much about docker and it would be nice to test it without setting up the databases on my local machine.

Alas, I have 2 problems with this. Although postgres works as expected I cannot get the mysql DB to work properly.

1) It appears that sometimes when I run `docker-compose up` it tries to run the tests before the mysql db is ready.
2) When the mysql db is ready I get a 
```
mysql_1      | 2017-01-29T19:18:56.464157Z 3 [Note] Access denied for user 'mysqlu'@'172.18.0.9' (using password: YES)
web_1        | django.db.utils.OperationalError: (1045, "Access denied for user 'mysqlu'@'172.18.0.9' (using password: YES)")
```

and yes if I attach to the mysql machine, I can connect to that db. Will have to figure out what it is I don't know.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.6
```
Comment (dlanger) [inline Dockerfile]: `447253099639.dkr.ecr.us-east-1.amazonaws.com/python:3.6.0-alpine--2016-12-28`, please.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    command: python runtests.py
+    environment:
+      - DATABASE_URL=postgres://postgres:postgres@postgres1:5432/postgres
+      - SHARD_001_DATABASE_URL=mysql://mysqlu:mysqlpw:@mysql:3306/mysqldb
+      - SHARD_002_DATABASE_URL=postgres://postgres:postgres@postgres2:5432/postgres
+      - SHARD_003_DATABASE_URL=postgres://postgres:postgres@postgres3:5432/postgres
+      - SHARD_004_DATABASE_URL=postgres://postgres:postgres@postgres4:5432/postgres
+      - REPLICA_001_DATABASE_URL=postgres://postgres:postgres@postgres5:5432/postgres
+      - REPLICA_002_DATABASE_URL=postgres://postgres:postgres@postgres6:5432/postgres
+    volumes:
```
Comment (dlanger) [inline docker-compose.yml]: Don't need this - the `ADD` should handle it for you.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    links:
+      - postgres1
+      - postgres2
+      - postgres3
+      - postgres4
+      - postgres5
+      - postgres6
+      - mysql
+
+  postgres1:
```
Comment (dlanger) [inline docker-compose.yml]: This spins up a bunch of individual `postgres` containers - is that what you were expecting?

**Comment by dlanger** [comment]
#1 is a known issue - Docker only guarantees that the other containers have started, not that they're necessarily ready to accept connections or anything. We're almost done https://github.com/waveaccounting/gottardo-nginx/pull/27 which should handle this for you, if you're willing to have some `nginx` sitting in your test container (or we can build one that has this, but no `nginx`). If not, you'll need to do something with [`wait-for-it.sh`](https://github.com/waveaccounting/gottardo-nginx/blob/master/wait-for-it.sh) (or similar) to block until it can make a TCP connection to the DBs and confirm they're ready.

File: docker-compose.yml (hunk 3 above)
Comment (JBKahn) [inline docker-compose.yml]: yes, although I suppose I could add a bunch of DBs to the same container. This seemed like an easy way to get started.

File: docker-compose.yml (hunk 2 above)
Comment (JBKahn) [inline docker-compose.yml]: oh!

File: Dockerfile (hunk 1 above)
Comment (JBKahn) [inline Dockerfile]: will change

File: docker-compose.yml (hunk 3 above)
Comment (dlanger) [inline docker-compose.yml]: Either works - a bunch of individual containers may be a little slower and will def have higher system requriements, but it'll do the job.

File: Dockerfile (hunk 1 above)
Comment (JBKahn) [inline Dockerfile]: it wasn't happy about my lack of auth, will `python:3.6.0-alpine` do?

**Comment by dlanger** [comment]
For #2, I bet it's something like the `mysqld.cnf` that ships with the container only allows local connections for [all users || the user `docker-entrypoint.sh` creates], or the user that's created has an IP restriction on it. I'd start with getting into the container and looking at `mysqld.cnf` and the permissions on the user created.

File: Dockerfile (hunk 1 above)
Comment (dlanger) [inline Dockerfile]: That container I linked is a Wave internal one, so you'd need to `make start` to log in.

For a public project `python:3.6.0-alpine` is fine, but you'll need to install the `pg` drivers before you install requirements - add this to your `Dockerfile`:
```
RUN apk add --no-cache --virtual .pg-drivers \
    postgresql-client \
    postgresql-dev
```
You'll have to do similar for MySQL too - [thi](https://pkgs.alpinelinux.org/packages) is the best way to figure out package names, or check out one of our apps that we've converted which use MySQL.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
+FROM python:3.6.0-alpine
```
Comment (JBKahn) [inline Dockerfile]:
```suggestion
Things
```

**Pull request closed without merging**